### ISSUE TITLE
Constant expressions are not optimized

### ISSUE THREAD

**Issue by Zylann** [godotengine/godot#7047]
Tested on Godot 2.1, Windows 10 64 bits

Q&A thread: https://godotengine.org/qa/9647/calculation-of-local-variable-inside-a-function

I tried to add the following test to my GDScriptPerformance project, to verify how long it takes for Godot to execute the addition of constants. Usually, compilers optimize that so if the expression is completely constant it would be resolved as an assignment.
However, Godot doesn't seem to do that, even in release builds.

Results of the release build:
```json
{
  "name": "Add two constants",
  "time": 23
},
{
  "name": "Add 20 constants",
  "time": 267
}
```

Adding 20 constants is 10 times slower than adding 2 constants.

Tests performed (TODO add them to https://github.com/godotengine/godot-tests/blob/master/benchmarks/zylann/gdscript_microtests.gd):
```gdscript
#-------------------------------------------------------
const CONST01 = 0
const CONST02 = 0

func test_constants_addition():
	start("Add two constants")
	for i in range(0,ITERATIONS):
		var x = CONST01 + CONST02
	
	stop()

#-------------------------------------------------------
const CONST03 = 0
const CONST04 = 0
const CONST05 = 0
const CONST06 = 0
const CONST07 = 0
const CONST08 = 0
const CONST09 = 0
const CONST10 = 0
const CONST11 = 0
const CONST12 = 0
const CONST13 = 0
const CONST14 = 0
const CONST15 = 0
const CONST16 = 0
const CONST17 = 0
const CONST18 = 0
const CONST19 = 0
const CONST20 = 0

func test_constants_addition_extended():
	start("Add 20 constants")
	for i in range(0,ITERATIONS):
		var x = CONST01 + CONST02 + CONST03 + CONST04 + CONST05 + CONST06 + CONST07 + CONST08 + CONST09 + CONST10 + \
		        CONST11 + CONST12 + CONST13 + CONST14 + CONST15 + CONST16 + CONST17 + CONST18 + CONST19 + CONST20
	
	stop()
```

**Comment by reduz** [comment]
not a chance, it's a dynamically typed language, just don't expect that
kind of optimizations

**Comment by Zylann** [comment]
Why do we have `const` then? I don't think it's a bug anyways, but I know there have been some discussion about const resolution, so this surprised me a bit. And I think doing this optimization is possible even in a dynamically typed language (just because we have `const`, really).

**Comment by vnen** [comment]
It's certain not *easy* to do, it would require changes in the compilation process. But I don't think it is impossible. Conceptually, you can detect that all constants are numbers and do the math at compile time, replacing the expression by the resulting value.

**Comment by reduz** [comment]
I think this is actually kind of being done, what is not is the loop
optimization

**Comment by Zylann** [comment]
I would not expect the loop to be optimized tbh^^

**Comment by vnen** [comment]
I think this still a problem for what I see in current GDScript code that I'm messing with. Expressions with literals are optimized, but the ones with identifiers are not (even when representing a constant).

**Comment by akien-mga** [comment]
Fixed by #40598 as per https://github.com/godotengine/godot/pull/38132#issuecomment-663883594.
